sbi/inference/posteriors/sbi_posterior.py

```diff
@@ -79,38 +82,65 @@ def log_prob(
         correct_for_leakage = normalize_snpe and self._alg_family == "snpe"
 
         # format inputs and context into the correct shape
-        inputs, context = utils.build_inputs_and_contexts(
+        inputs, context = utils.match_shapes_of_inputs_and_contexts(
             inputs, context, self._context, correct_for_leakage
         )
 
         # go into evaluation mode
         self.neural_net.eval()
 
         # compute the unnormalized log probability by evaluating the network
-        unnormalized_log_prob = self.neural_net.log_prob(inputs, context)
+        if self._alg_family == "snpe":
+            unnormalized_log_prob = self.neural_net.log_prob(inputs, context)
+        elif self._alg_family == "sre":
+            warn(
+                "The log-probability returned by SRE is only correct up to a "
+                "normalizing constant."
+            )
+            log_ratio = self.neural_net(torch.cat((inputs, context)).reshape(1, -1))
+            unnormalized_log_prob = log_ratio + self._prior.log_prob(inputs)
+        else:
+            raise NameError(
+                "Evaluating the log-probability can only be done for SNPE and SRE."
+            )
+
+        if correct_for_leakage:
+            # set the log-likelihood to -infinity if parameter set (inputs) is outside
+            # of prior bounds.
+            prior_log_probs = self._prior.log_prob(inputs)
+            unnormalized_log_prob[torch.isinf(prior_log_probs)] = torch.tensor(
```

Comment:
There's a helpful idiom for conditional substitution of values in a tensor, using `torch.where`; you can find a relevant example (same goal of zeroing-away probability outside prior) here: https://github.com/mackelab/sbi/blob/cb945c338b99faaef609afe49469d9a1f15aaf0c/sbi/inference/snpe/snpe_base.py#L537.

Reply:
This is pretty cool, thanks!

Reply:
is this changed to `pytorch.where` now?

Reply:
This will be changed in the next commit.